### PR TITLE
Create CAPI objects from EKS-A spec

### PR DESCRIPTION
**Pull request by mdsgabriel**
*Issue #, if available:*

*Description of changes:*
This will create CAPI objects form a EKS-Anywhere spec and apply them. 
Next steps will be applying extra objects and the CNI plugin.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: config/rbac/manager_extraroles_patch.yaml [hunk 1]
```diff
@@ -2,11 +2,11 @@
   path: /rules/-
   value:
     apiGroups:
-      - cluster.x-k8s.io
+      - etcdcluster.cluster.x-k8s.io
```
Comment (jiayiwang7) [inline config/rbac/manager_extraroles_patch.yaml:5]: are `etcdcluster.cluster.x-k8s.io` and `controlplane.cluster.x-k8s.io`  still needed when u have `cluster.x-k8s.io` defined?

File: controllers/controllers/clusters/vsphere.go [hunk 2]
```diff
@@ -135,5 +137,49 @@ func (v *VSphereClusterReconciler) Reconcile(ctx context.Context, cluster *anywh
 		return reconciler.Result{}, err
 	}
 
+	workerNodeGroupMachineSpecs := make(map[string]anywherev1.VSphereMachineConfigSpec, len(machineConfigMap))
```
Comment (jiayiwang7) [inline controllers/controllers/clusters/vsphere.go:140]: should it be `len(machineConfigMap)-1` ? :0

File: controllers/controllers/clusters/vsphere.go [hunk 3]
```diff
@@ -135,5 +137,49 @@ func (v *VSphereClusterReconciler) Reconcile(ctx context.Context, cluster *anywh
...
+	for _, wnConfig := range cluster.Spec.WorkerNodeGroupConfigurations {
+		workerNodeGroupMachineSpecs[wnConfig.MachineGroupRef.Name] = machineConfigMap[wnConfig.MachineGroupRef.Name].Spec
+	}
+
+	cp := machineConfigMap[specWithBundles.Spec.ControlPlaneConfiguration.MachineGroupRef.Name]
+	var etcd *anywherev1.VSphereMachineConfigSpec
+	if specWithBundles.Spec.ExternalEtcdConfiguration != nil {
+		etcdSpec := machineConfigMap[specWithBundles.Spec.ExternalEtcdConfiguration.MachineGroupRef.Name]
```
Comment (jiayiwang7) [inline controllers/controllers/clusters/vsphere.go:148]: is the naming up side down?

```suggestion
etcd := machineConfigMap[specWithBundles.Spec.ExternalEtcdConfiguration.MachineGroupRef.Name]
etcdSpec = &etcd.Spec
```

File: controllers/controllers/clusters/vsphere.go [hunk 4]
```diff
@@ -135,5 +137,49 @@ func (v *VSphereClusterReconciler) Reconcile(ctx context.Context, cluster *anywh
...
+	}
+
+	templateBuilder := vsphere.NewVsphereTemplateBuilder(&dataCenterConfig.Spec, &cp.Spec, etcd, workerNodeGroupMachineSpecs, time.Now, true)
+	clusterName := cluster.ObjectMeta.Name
+
+	cpOpt := func(values map[string]interface{}) {
+		values["controlPlaneTemplateName"] = templateBuilder.CPMachineTemplateName(clusterName)
+		controlPlaneUser := machineConfigMap[cluster.Spec.ControlPlaneConfiguration.MachineGroupRef.Name].Spec.Users[0]
+		values["vsphereControlPlaneSshAuthorizedKey"] = controlPlaneUser.SshAuthorizedKeys[0]
+		etcdUser := machineConfigMap[cluster.Spec.ExternalEtcdConfiguration.MachineGroupRef.Name].Spec.Users[0]
```
Comment (jiayiwang7) [inline controllers/controllers/clusters/vsphere.go:159]: what if `ExternalEtcdConfiguration` does not exist? (stacked etcd)

File: controllers/controllers/clusters/vsphere.go [hunk 5]
```diff
@@ -135,5 +137,49 @@ func (v *VSphereClusterReconciler) Reconcile(ctx context.Context, cluster *anywh
...
+
+	workersSpec, err := templateBuilder.GenerateCAPISpecWorkers(specWithBundles, nil)
+	if err != nil {
+		return reconciler.Result{}, err
+	}
+
+	if err := reconciler.ReconcileYaml(ctx, v.Client, controlPlaneSpec); err != nil {
+		return reconciler.Result{Result: &ctrl.Result{
+			Requeue:      true,
+			RequeueAfter: time.Minute,
```
Comment (jiayiwang7) [inline controllers/controllers/clusters/vsphere.go:177]: can we maybe use `const` for requeue time?

File: controllers/controllers/clusters/vsphere.go [hunk 6]
```diff
@@ -135,5 +137,49 @@ func (v *VSphereClusterReconciler) Reconcile(ctx context.Context, cluster *anywh
...
+		}}, err
+	}
+	if err := reconciler.ReconcileYaml(ctx, v.Client, workersSpec); err != nil {
+		return reconciler.Result{}, err
```
Comment (jiayiwang7) [inline controllers/controllers/clusters/vsphere.go:181]: we should also requeue here right?

File: controllers/controllers/clusters/vsphere.go (hunk 2 above)
Comment (mdsgabriel) [reply controllers/controllers/clusters/vsphere.go]: using len(cluster.Spec.WorkerNodeGroupConfigurations)

File: config/rbac/manager_extraroles_patch.yaml (hunk 1 above)
Comment (mdsgabriel) [reply config/rbac/manager_extraroles_patch.yaml:5]: yes

File: controllers/controllers/clusters/vsphere.go (hunk 3 above)
Comment (mdsgabriel) [reply controllers/controllers/clusters/vsphere.go]: done

File: controllers/controllers/clusters/vsphere.go (hunk 4 above)
Comment (mdsgabriel) [reply controllers/controllers/clusters/vsphere.go]: added a check

File: controllers/controllers/clusters/vsphere.go (hunk 5 above)
Comment (mdsgabriel) [reply controllers/controllers/clusters/vsphere.go]: added a constant, but this is a temporary code and will probably change

**Review by jiayiwang7 — APPROVED** [review]

**Pull request merged**